### ISSUE TITLE
Any way to query the value of "Label" property of AXUIElemnt?

### ISSUE THREAD

**Issue by avegetablechicken** [Hammerspoon/hammerspoon#3713]
macOS tends to write more and more useful values as  "Label" property instead of "Title" or "Value" properties of UIElements. I found no methods to query the value of "Label" property either by AppleScript or by `hs.axuielement`. Anybody knows way to workaround it?

**Comment by latenitefilms** [comment]
You should be able to access every AX element of an object. Use Xcode's Accessibility Inspector to see what values you have access to.

CommandPost has a bunch of AX helper utilities that might be helpful?

https://github.com/CommandPost/CommandPost/blob/develop/src/extensions/cp/ui/axutils.lua

**Comment by avegetablechicken** [comment]
@latenitefilms It seems that all properties accessible in Accessibility Inspector can be accessed by `hs.axuielement` except "Label" :(

**Comment by latenitefilms** [comment]
Are you able to give an example of a built-in macOS application that uses this label so I can try and reproduce?

Can you take a screenshot of what you're seeing in Accessibility Inspector, or better yet, UI Browser (with "Technical Terminology" turned on)?

Thoughts @asmagill ?

**Comment by asmagill** [comment]
"AXLabelUIElements" and "AXLabelValue" are defined for URLs (i.e. disks, files, and network connections), but are not considered standard for other objects. That said, an application can define any attributes it wants to, so its entirely possible that its a valid property for some applications.

I agree that we need to know more -- what applications are being queried, screen shots of what UI Browser is showing, what does `hs.inspect(hs.axuielement.applicationElement(hs.application("app-name")): attributeNames())` (or equivalent), etc.

**Comment by asmagill** [comment]
Oh... Label not Title... misremembered. Same answer, though... a lot of `error = "Requested value does not exist"`.

**Comment by latenitefilms** [comment]
"Requested value does not exist" wouldn't be that unexpected would it? I feel like most AX values aren't actually populated?

**Comment by latenitefilms** [comment]
<img width="1109" alt="Screenshot 2024-11-11 at 12 01 50 PM" src="https://github.com/user-attachments/assets/9f78ec86-7967-40d1-9556-0fde844ea7f2">

**Comment by latenitefilms** [comment]
I'm not sure where Accessibility Inspector actually gets the `Label` - I'm GUESSING it's probably actually using `AXValue`?

**Comment by asmagill** [comment]
> "Requested value does not exist" wouldn't be that unexpected would it? I feel like most AX values aren't actually populated?

Yes and no... the code snippit I gave decides what attributes to query based on the results from `hs.axuielement:attributeNames()` which returns a table of attribute names that the application *claims* to provide. There is no actual check by Apple or the OS that the app actually does provide those attributes, and the return value for the attribute names *is* determined by the programmers, so... it could be that they copy-pasted and forgot/didn't intend to include that property, or it could be that the application performed some of its own internal logic and decided not to provide an answer at this time.

**Comment by asmagill** [comment]
If you run the snippit I gave above, it will dump the Control Center applications properties to the console, and AXLabel does appear for some items in the returned output; it just returns an error when the property is actually queried (by Hammerspoon at least)
